go_covert_lib/controller/channel/httpNormal/httpNormal.go

```diff
@@ -0,0 +1,284 @@
+			w.Header().Add("Valid", "1")
+		}
+
+		// if the request message is a post request
+		// extract the data from the body of the request message and prepare it to be displayed to the user
+	} else if r.Method == "POST" {
+		buf := bytes.Buffer{}
+		buf.ReadFrom(r.Body)
+		select {
+		case c.serverRecBuf <- buf.Bytes():
+		case <-time.After(time.Millisecond * 50):
+			//log.Println("POST Request timeout")
+		case <-c.cancel:
+		}
+	} else {
+		log.Println("Invalid Covert Message")
+	}
+}
+
+// Close the covert the channel
+func (c *Channel) Close() error {
+	select {
+	case <-c.cancel:
+		return nil
+	default:
+		close(c.cancel)
+		if c.conf.UserType == Server {
+			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel()
+			return c.srv.Shutdown(ctx)
+		}
+	}
+	return nil
+}
+
+//receive information from another computer in the form of data in a byte array
+func (c *Channel) Receive(data []byte) (uint64, error) {
+
+	// if the user is a server type computer
+	if c.conf.UserType == Server {
+
+		//Determines whether the channel is handling timeout or not
+		//then read from the servers receive buffer
+		//this is without timeout
+		if c.conf.ReadTimeout == 0 {
+			select {
+			case httpdata := <-c.serverRecBuf:
+				copy(data, httpdata)
+				if len(httpdata) > len(data) {
+					return uint64(len(data)), errors.New("Buffer overflow")
+				} else {
+					return uint64(len(httpdata)), nil
+				}
+			case <-c.cancel:
+				return 0, errors.New("Channel closed")
+			}
+			// this is with timeout
+		} else {
+			select {
+			case httpdata := <-c.serverRecBuf:
+				copy(data, httpdata)
+				if len(httpdata) > len(data) {
+					return uint64(len(data)), errors.New("Buffer overflow")
+				} else {
+					return uint64(len(httpdata)), nil
+				}
+			case <-time.After(c.conf.WriteTimeout):
+				return 0, errors.New("Write Timeout")
+			case <-c.cancel:
+				return 0, errors.New("Channel closed")
+			}
+		}
+
+		// if the user is client type computer
+	} else {
+		n, success, err := c.clientRequest(data)
+		if success || err != nil {
+			return n, err
+		} else {
+			if c.conf.ClientPollRate > 0 {
+				// Loop while making requests
+				var ticker *time.Ticker = time.NewTicker(c.conf.ClientPollRate)
+				defer ticker.Stop()
+				if c.conf.ClientTimeout == time.Duration(0) {
+					// No timeout
+					for {
+						select {
+						case <-ticker.C:
+							n, success, err := c.clientRequest(data)
+							if success || err != nil {
+								return n, err
+							}
+						case <-c.cancel:
+							return 0, errors.New("Channel closed")
+						}
+					}
+				} else {
+					for {
+						select {
+						case <-ticker.C:
+							n, success, err := c.clientRequest(data)
+							if success || err != nil {
+								return n, err
+							}
+						case <-time.After(c.conf.ClientTimeout):
+							return 0, errors.New("Client Timeout")
+						case <-c.cancel:
+							return 0, errors.New("Channel closed")
+						}
+					}
+				}
+			} else {
```

Comment:
As mentioned above, you have to manually create an error here

Reply:
resolved